core/src/main/scala/kafka/coordinator/TransactionLogManager.scala

```diff
@@ -17,15 +17,235 @@
+
+  private val VALUE_SCHEMA_V0 = new Schema(new Field(PID_KEY, INT64),
+                                           new Field(EPOCH_KEY, INT16),
+                                           new Field(TXN_TIMEOUT_KEY, INT32),
+                                           new Field(TXN_STATUS_KEY, INT8),
+                                           new Field(TXN_PARTITIONS_KEY, new ArrayOf(VALUE_PARTITIONS_SCHEMA)) )
+  private val VALUE_SCHEMA_PID_FIELD = VALUE_SCHEMA_V0.get(PID_KEY)
+  private val VALUE_SCHEMA_EPOCH_FIELD = VALUE_SCHEMA_V0.get(EPOCH_KEY)
+  private val VALUE_SCHEMA_TXN_TIMEOUT_FIELD = VALUE_SCHEMA_V0.get(TXN_TIMEOUT_KEY)
+  private val VALUE_SCHEMA_TXN_STATUS_FIELD = VALUE_SCHEMA_V0.get(TXN_STATUS_KEY)
+  private val VALUE_SCHEMA_TXN_PARTITIONS_FIELD = VALUE_SCHEMA_V0.get(TXN_PARTITIONS_KEY)
+
+  private val KEY_SCHEMAS = Map(
+    0 -> KEY_SCHEMA_V0)
+
+  private val VALUE_SCHEMAS = Map(
+    0 -> VALUE_SCHEMA_V0)
+
+  private val CURRENT_KEY_SCHEMA_VERSION = 0.toShort
+  private val CURRENT_VALUE_SCHEMA_VERSION = 0.toShort
+
+  private val CURRENT_KEY_SCHEMA = schemaForKey(CURRENT_KEY_SCHEMA_VERSION)
+
+  private val CURRENT_VALUE_SCHEMA = schemaForValue(CURRENT_VALUE_SCHEMA_VERSION)
+
+  private def schemaForKey(version: Int) = {
+    val schemaOpt = KEY_SCHEMAS.get(version)
+    schemaOpt match {
+      case Some(schema) => schema
+      case _ => throw new KafkaException(s"Unknown transaction log message key schema version $version")
+    }
+  }
+
+  private def schemaForValue(version: Int) = {
+    val schemaOpt = VALUE_SCHEMAS.get(version)
+    schemaOpt match {
+      case Some(schema) => schema
+      case _ => throw new KafkaException(s"Unknown transaction log message value schema version $version")
+    }
+  }
+
+  /**
+    * Generates the bytes for transaction log message key
+    *
+    * @return key bytes
+    */
+  private[coordinator] def keyToBytes(transactionalId: String): Array[Byte] = {
+    val key = new Struct(CURRENT_KEY_SCHEMA)
+    key.set(KEY_SCHEMA_TXN_ID_FIELD, transactionalId)
+
+    val byteBuffer = ByteBuffer.allocate(2 /* version */ + key.sizeOf)
+    byteBuffer.putShort(CURRENT_KEY_SCHEMA_VERSION)
+    key.writeTo(byteBuffer)
+    byteBuffer.array()
+  }
+
+  /**
+    * Generates the payload bytes for transaction log message value
+    *
+    * @return value payload bytes
+    */
+  private[coordinator] def valueToBytes(pidMetadata: PidMetadata): Array[Byte] = {
+    val value = new Struct(CURRENT_VALUE_SCHEMA)
+    value.set(VALUE_SCHEMA_PID_FIELD, pidMetadata.pid)
+    value.set(VALUE_SCHEMA_EPOCH_FIELD, pidMetadata.epoch)
+    value.set(VALUE_SCHEMA_TXN_TIMEOUT_FIELD, pidMetadata.txnTimeoutMs)
+    value.set(VALUE_SCHEMA_TXN_STATUS_FIELD, pidMetadata.txnMetadata.state.byte)
+
+    if (pidMetadata.txnMetadata.state.equals(NotExist)) {
+      if (pidMetadata.txnMetadata.topicPartitions.nonEmpty)
+        throw new IllegalStateException(s"Transaction status ${pidMetadata.txnMetadata} is not valid")
```

Comment:
This is a bit of a cryptic message. From the code, it looks like a transactioanalId -> PID message should not have any topic partitions or other transaction status associated with it. It would be good to spell this out in the error message.